### PR TITLE
add spectrum class

### PR DESCRIPTION
**Pull request by drammock**
Progress so far:

- [x] `.compute_psd()` method for Raw, Epochs, Evoked yields Spectrum object
- [x] supports method='welch' and method='multitaper'
- [x] supports unaggregated welch via `average=False`
- [x] supports unaggregated multitaper via `output='complex'`
- [ ] ~~support passing a callable to aggregate welch/multitaper estimates~~ #11093
- [x] properties ch_names, freqs, method (welch/multitaper)
- [x] private property `._dims`, e.g., `('epoch', 'channel', 'freq', 'taper')` for an unaggregated multitaper derived from Epochs
- [x] save to / load from HDF5 (`spectrum.save(fname)` / `mne.time_frequency.read_spectrum(fname)`) for Spectrum object
- [x] save/load for EpochsSpectrum object
- [x] `__eq__` method for comparisons / testing
- [x] `__getitem__` works for raw/epochs/evoked (raw/evoked-derived spectra work like np.array; epochs-derived spectra work like epochs: select by condition name, pandas query, etc)
- [x] `.pick()` method
- [x] `.units()` method
- [x] `.to_data_frame()` method (works with raw/epochs/evoked, agged/unagged estimates)
- [x] `.plot()` method with equivalent API to old `inst.plot_psd()` method
- [x] `.plot_topomap()` method
- [x] `.plot_topo()` method
- [x] tests
- [x] tutorial
- [x] deprecations
- [x] changelog

```py
import os
import mne

sample_data_folder = mne.datasets.sample.data_path()
sample_data_raw_file = os.path.join(sample_data_folder, 'MEG', 'sample',
                                    'sample_audvis_raw.fif')
raw = mne.io.read_raw_fif(sample_data_raw_file, verbose=False, preload=False)

spectrum = raw.compute_psd()
spectrum.plot()
```
yields this:
Figure_1

`spectrum` has properties `_data`, `freqs`, `ch_names`, `method` (welch or multitaper), an `info` dict, and `spec._dims` (description of the data dims). For example

```py
In [8]: unagg_spec = raw.to_spectrum(average=None)
In [9]: unagg_spec._dims
Out[9]: ('channel', 'freq', 'segment')
```
Here is the corresponding repr `<Spectrum (from Raw) | 364 channels × 129 freqs × 651 segments, 0.0-300.3 Hz>`

~~Eventually (in another PR) I'd like to refactor the multitaper functions to make it possible to return the unaggregated tapers~~ **EDIT** *I think this was done already while I was on leave*, in which case you'd expect something like `raw.to_spectrum(multitaper, aggregate=False)._dims` to be `('channel', 'freq', 'taper')`.

`epochs.compute_psd()` and `evoked.compute_psd()` also work:

```py
events = mne.find_events(raw, stim_channel='STI 014')
event_dict = {'auditory/left': 1, 'auditory/right': 2, 'visual/left': 3}
epochs = mne.Epochs(raw, events, event_id=event_dict, preload=False,
                    proj=False)
evk = epochs.average()

epo_spec = epochs.compute_psd()
evk_spec = evk.compute_psd()
```

### CODE REVIEW

**Comment by drammock** [comment]
this is still a draft but now would be a good time for some early feedback about code organization and API (ping @larsoner and @agramfort in particular, but input is welcome from anyone).

**Comment by drammock** [comment]
> I am wondering about the use of spectrum. We use psd everywhere else.

I chose "spectrum" because it's more general: e.g., depending on what parameters people use, sometimes we plot the amplitude *spectrum*, or we plot 10*log10 of the power *spectrum*.

**Comment by drammock** [comment]
Other reasons to pick a new name:

- people are used to things like `psds, freqs = psd_welch(raw)` so maybe (?) `raw.compute_psd()` might be similar enough to generate wrong expectations about the return value?
- having a more distinct name makes it less surprising that there are API changes.  For example the previous psd plotting functions have `area_mode` parameter which I'm proposing to change to `ci` in `spectrum.plot()` (`ci` is used in seaborn, and also in our own `plot_compare_evokeds`).

**Comment by agramfort** [comment]
ok makes sense

**Comment by dengemann** [comment]
@dengemann tagging myself to follow. Great this is alive!

**Comment by dengemann** [comment]
Really happy to see this @drammock. Once basics are settled it would be cool to also add `plot_topomap`, based on `Epochs.plot_psd_topomap`. Not to forget adjusted `.to_data_frame`. Not suggesting all this should happen in this PR though :)

File: mne/time_frequency/tests/test_spectrum.py
Comment (drammock) [inline mne/time_frequency/tests/test_spectrum.py:56]: @larsoner this test fails for evoked because of a limitation in pytest. The failure is:

```
_____________________________________________ test_spectrum_io[evoked] ______________________________________________
The requested fixture has no parameter defined for test:
    mne/time_frequency/tests/test_spectrum.py::test_spectrum_io[evoked]

Requested fixture '_evoked' defined in:
mne/conftest.py:306

Requested here:
/opt/miniforge3/envs/mnedev/lib/python3.10/site-packages/_pytest/fixtures.py:659
```

I can make it go away by removing `params=[testing._pytest_param()]` from the signature here:

https://github.com/mne-tools/mne-python/blob/2ba483ab91828be18a480e689d95f61c422ed6eb/mne/conftest.py#L305-L312

But presumably that will break your magic auto-skipping behavior.  The limitation is that `request.getfixturevalue()` cannot handle fixtures that have params, and this is [apparently by design](https://github.com/pytest-dev/pytest/issues/4666#issuecomment-456593913). A workaround is to use the [pytest-lazy-fixture](https://pypi.org/project/pytest-lazy-fixture) package.  Another alternative is to copy-paste a separate test for Evoked IO, like this:

```py
@requires_h5py
def test_evoked_spectrum_io(evoked, tmp_path):
    """Test save/load of spectrum objects."""
    fname = tmp_path / 'spectrum.h5'
    orig = evoked.compute_psd()
    orig.save(fname)
    loaded = read_spectrum(fname)
    assert orig == loaded
```
... which works because now the `evoked` fixture is listed directly in the test signature (instead of being accessed via `getfixturevalue()`), but is not very DRY.

What do you recommend?

Comment (larsoner) [reply mne/time_frequency/tests/test_spectrum.py]: I would work around it with:
```
@requires_h5py
@pytest.mark.parametrize('inst', ('raw', 'epochs', 'evoked'))
def test_spectrum_io(inst, tmp_path, request, evoked):
    """Test save/load of spectrum objects."""
    # work around parametrized fixtures not being accessible
    # via request.getfixturevalue
    # https://github.com/pytest-dev/pytest/issues/4666#issuecomment-456593913  # noqa: E501
    if inst == 'evoked':
        inst = evoked
    else:
        inst = request.getfixturevalue(inst)
    fname = tmp_path / 'spectrum.h5'
    orig = inst.compute_psd()
    orig.save(fname)
    loaded = read_spectrum(fname)
    assert orig == loaded
```

Comment (drammock) [reply mne/time_frequency/tests/test_spectrum.py]: Ah, OK. Inelegant but practical. Will do.

**Comment by larsoner** [comment]
@drammock the bullet point is still unchecked:

>  support passing a callable to aggregate welch/multitaper estimates

Should we wait for a follow-up PR, or did you want to do it here?

Either way I can look at the code early tomorrow

**Comment by drammock** [comment]
That was added to the list this week after conversations with @ajquinn. I think a follow up PR is probably best but if you disagree I can add it here. I don't think it will be too hard.

-------- Original Message --------

File: mne/viz/epochs.py
Comment (larsoner) [inline mne/viz/epochs.py:993]: Why not just pass `show=show` above, then you don't need `plt_show(show, fig)` here?

File: mne/viz/raw.py
Comment (larsoner) [inline mne/viz/raw.py:426]: Same here and maybe elsewhere

File: mne/viz/utils.py
Comment (larsoner) [inline mne/viz/utils.py:2251]: Shouldn't you need to pass the `area_mode` value to `_ci` ? 
```suggestion
            else:  # area_mode is float
                hyp_limits = _ci(psd, area_mode)
```

File: tutorials/epochs/20_visualize_epochs.py
Comment (larsoner) [inline tutorials/epochs/20_visualize_epochs.py:163]: It is weird to me to say this, then immediately use such legacy code below. Let's just switch to the new API below, too (and elsewhere). I think for legacy stuff where we keep around 2-line wrappers in order to avoid breaking people's code we can do this sort of thing.

File: tutorials/raw/40_visualize_raw.py
Comment (larsoner) [inline tutorials/raw/40_visualize_raw.py:134]: This one by contrast seems fine because the old API is *not* used immediately afterward

File: mne/time_frequency/spectrum.py
Comment (larsoner) [inline mne/time_frequency/spectrum.py:371]: Have you tried `'<br>'.join()` or something similar? I think it would be more readable and less risk of becoming really wide for a single row's data in 

https://output.circle-artifacts.com/output/job/b0cf67fb-c753-4c24-9ec2-0e9877d45f9e/artifacts/0/dev/auto_tutorials/time-freq/10_spectrum_class.html

File: tutorials/time-freq/10_spectrum_class.py
Comment (larsoner) [inline tutorials/time-freq/10_spectrum_class.py:135]: The table title is off-center for me

https://output.circle-artifacts.com/output/job/b0cf67fb-c753-4c24-9ec2-0e9877d45f9e/artifacts/0/dev/auto_tutorials/time-freq/10_spectrum_class.html#id4

Screen Shot 2022-08-25 at 9 29 24 AM


Can you either convert this to an issue, see if the fix is easy, or open an upstream pydata-sphinx-theme issue?

File: mne/time_frequency/spectrum.py
Comment (larsoner) [inline mne/time_frequency/spectrum.py:294]: We typically use `_get_args` from `fixes` for this. Can you use it here and/or open an issue that we should investigate whether we can migrate to `inspect.signature` directly nowadays? I can't remember if we just needed to wait until 3.7+ or if `.parameters` isn't enough sometimes, or something else

File: mne/time_frequency/spectrum.py
Comment (larsoner) [inline mne/time_frequency/spectrum.py:338]: Can you make a comment how this differs from the default implementation / why it's needed? To me it looks like the default and I'm a bit surprised you need it here at all

File: mne/time_frequency/spectrum.py
Comment (larsoner) [inline mne/time_frequency/spectrum.py:352]: Maybe a better approach would be to make `object_diff` handle classes? I think in principle something like this could work:
```
if inspect.isclass(a):
    if not inspect.isclass(b):
        same = False
    else:
        same = (a == b)
```

**Review by larsoner — COMMENTED** [review]
Very clean! Just some minor comments from me, otherwise LGTM +1 for merge

File: tutorials/epochs/20_visualize_epochs.py
Comment (drammock) [reply tutorials/epochs/20_visualize_epochs.py:163]: haha, I didn't mean to leave the old API in the tutorial. thanks for catching that.

File: tutorials/time-freq/10_spectrum_class.py
Comment (drammock) [reply tutorials/time-freq/10_spectrum_class.py:135]: it scrolls.

File: mne/viz/epochs.py
Comment (drammock) [reply mne/viz/epochs.py:993]: good catch. actually already passing `show=show` above so this can just be removed.

File: mne/time_frequency/spectrum.py
Comment (drammock) [reply mne/time_frequency/spectrum.py:371]: I had the same thought but `'<br/>'.join` did not work (the tag was rendered in the output).  I'll do a bit of searching to see if I can figure out another way to add linebreaks

File: mne/time_frequency/spectrum.py
Comment (drammock) [reply mne/time_frequency/spectrum.py:338]: originally I thought I had to override the one inherited from Epochs, but whatever the problem was it seems to be gone now.

File: mne/time_frequency/spectrum.py
Comment (drammock) [reply mne/time_frequency/spectrum.py:294]: #11089

File: doc/references.bib
Comment (wmvanvliet) [inline doc/references.bib:2290]:
```suggestion
```

File: mne/viz/raw.py
Comment (wmvanvliet) [inline mne/viz/raw.py:378]:
```suggestion
        Number of points to use in Welch FFT calculations. Default is ``None``,
```

File: mne/viz/topomap.py
Comment (wmvanvliet) [inline mne/viz/topomap.py:1949]: Why are these commented-out?

File: mne/viz/topomap.py
Comment (wmvanvliet) [inline mne/viz/topomap.py:2010]: This looks like some TODO but you want to hold off on it until topomap plotting has improved?

**Comment by drammock** [comment]
- [new tutorial](https://output.circle-artifacts.com/output/job/e98a0c2d-c9ea-4c7c-adc5-f67ee86748aa/artifacts/0/dev/auto_tutorials/time-freq/10_spectrum_class.html#sphx-glr-auto-tutorials-time-freq-10-spectrum-class-py)
- changed tutorial [sleep stage classification](https://output.circle-artifacts.com/output/job/e98a0c2d-c9ea-4c7c-adc5-f67ee86748aa/artifacts/0/dev/auto_tutorials/clinical/60_sleep.html#sphx-glr-auto-tutorials-clinical-60-sleep-py)
- changed tutorial [raw plotting methods](https://output.circle-artifacts.com/output/job/e98a0c2d-c9ea-4c7c-adc5-f67ee86748aa/artifacts/0/dev/auto_tutorials/raw/40_visualize_raw.html#sphx-glr-auto-tutorials-raw-40-visualize-raw-py)
- changed tutorial [visualizing epochs](https://output.circle-artifacts.com/output/job/e98a0c2d-c9ea-4c7c-adc5-f67ee86748aa/artifacts/0/dev/auto_tutorials/epochs/20_visualize_epochs.html#sphx-glr-auto-tutorials-epochs-20-visualize-epochs-py)

File: tutorials/clinical/60_sleep.py
Comment (wmvanvliet) [inline tutorials/clinical/60_sleep.py:190]:
```suggestion
stages = sorted(event_id.keys())
```
Explicit is better than implicit.

File: mne/time_frequency/spectrum.py
Comment (wmvanvliet) [inline mne/time_frequency/spectrum.py:17]: These commented-out things need to be addressed somehow sometime.

**Review by wmvanvliet — COMMENTED** [review]
It's a work of art! 🤩 

Do we at some point want to support users creating their own `Spectral` instances like `EpochsArray` and `EvokedArray`?

**Comment by drammock** [comment]
> Do we at some point want to support users creating their own `Spectral` instances like `EpochsArray` and `EvokedArray`?

this is something I struggled with, and ultimately decided "YAGNI".  But, I suspect it wouldn't be too difficult given that it's already basically implemented in the `_from_file` method for loading from disk.

File: mne/decoding/transformer.py
Comment (drammock) [reply mne/decoding/transformer.py:15]: I don't understand why but it was at some point causing for me a circular import error.

File: mne/preprocessing/ica.py
Comment (drammock) [reply mne/preprocessing/ica.py:1651]: yes.

**Comment by wmvanvliet** [comment]
> ultimately decided "YAGNI"

Fair enough.

File: mne/time_frequency/tests/test_psd.py
Comment (drammock) [reply mne/time_frequency/tests/test_psd.py:96]: yes. These old tests operated on the `psd_welch` and `psd_multitaper` funcs (which are now deprecated); now these tests only use `psd_array_[welch|multitaper]` (which don't take `proj`).  BUT testing with `proj` either True or False is still done, it's just done in `tests/test_spectrum.py` now.

File: mne/utils/docs.py
Comment (drammock) [reply mne/utils/docs.py:1549]: nope. (if it was, CIs would have complained. Also I checked: `git grep "%(index_df)s"`)

File: mne/viz/topomap.py
Comment (drammock) [reply mne/viz/topomap.py:1949]: these will be uncommented in a subsequent PR, where ALL the plot_topomap methods (for Cov, Proj, etc) will get their APIs standardized.

File: mne/time_frequency/spectrum.py
Comment (drammock) [reply mne/time_frequency/spectrum.py:17]: these will be uncommented in a subsequent PR, where ALL the plot_topomap methods (for Cov, Proj, etc) will get their APIs standardized.

**Review by larsoner — APPROVED** [review]
LGTM +1 for merge after the `_ci` fix!

**Review by wmvanvliet — APPROVED** [review]

**Comment by larsoner** [comment]
https://output.circle-artifacts.com/output/job/79b51cb7-962f-4095-87ce-5aa72b49dcc1/artifacts/0/dev/auto_tutorials/time-freq/10_spectrum_class.html#sphx-glr-auto-tutorials-time-freq-10-spectrum-class-py

**Comment by larsoner** [comment]
Congrats @drammock !

**Comment by wmvanvliet** [comment]
Nice one!

**Comment by agramfort** [comment]
🎉 🍻 👏 @drammock

**Pull request merged**
